README.md
## cosmic_frontmatter

`cosmic_frontmatter` is a package that provides a simple way to parse the frontmatter of a markdown file.

## Getting started

To get started, you need to install the package by adding `cosmic_frontmatter` to your `pubspec.yaml`:

```yaml
dependencies:
  flutter:
    sdk: flutter

  cosmic_frontmatter: ^1.0.3
```

## Usage

`cosmic_frontmatter` exports a `parseFrontmatter` function that takes a string (your markdown document) and a `builder` function. The `builder` function is called with the parsed frontmatter as a map.

This allows you to get typed frontmatter. Here's an example that uses [json_serializable](https://developer.school/tutorials/flutter-using-json_serializable-to-serialise-dart-classes) and [freezed](https://developer.school/how-to-use-freezed-with-flutter) to generate a `MockFrontmatter` class:

```dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'mock_frontmatter.freezed.dart';
part 'mock_frontmatter.g.dart';

@freezed
class MockFrontmatter with _$MockFrontmatter {
  const factory MockFrontmatter({
    required String title,
    required String author,
    required String excerpt,
    required String category,
    required String date,
  }) = _MockFrontmatter;

  const MockFrontmatter._();

  factory MockFrontmatter.fromJson(Map<String, dynamic> json) => _$MockFrontmatterFromJson(json);
}
```

With the `MockFrontmatter` in place, you can parse the frontmatter of a markdown file:

```dart
const mockDocument = """
---
title: This is a test document
author: paul-halliday
excerpt: Learn how to use Markdown with Flutter at developer.school
category: Flutter
date: "2021-10-25"
---

You can learn how to use Markdown with Flutter at [developer.school](https://developer.school/tutorials/how-to-display-markdown-in-flutter).
""";

final mockFrontmatter = parseFrontmatter(
  content: mockDocument,
  frontmatterBuilder: (map) {
    // Do anything you want with the `map`:

    return MockFrontmatter.fromJson(map);
});
```

This can then be used in your app however you want. Here's an example which merely displays the frontmatter on the screen, along-side the markdown body:

```dart
class _MyHomePageState extends State<MyHomePage> {
  late Document<MockFrontmatter> document;

  @override
  void initState() {
    super.initState();

    document = parseFrontmatter(
      content: mockDocument,
      frontmatterBuilder: (map) => MockFrontmatter.fromJson(map),
    );
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(document.frontmatter.title),
      ),
      body: Padding(
        padding: const EdgeInsets.all(16.0),
        child: Center(
          child: Column(
            crossAxisAlignment: CrossAxisAlignment.stretch,
            children: [
              ListTile(
                contentPadding: EdgeInsets.zero,
                title: const Text('Title'),
                subtitle: Text(document.frontmatter.title),
              ),
              ListTile(
                contentPadding: EdgeInsets.zero,
                title: const Text('Author'),
                subtitle: Text(document.frontmatter.author),
              ),
              ListTile(
                contentPadding: EdgeInsets.zero,
                title: const Text('Excerpt'),
                subtitle: Text(document.frontmatter.excerpt),
              ),
              ListTile(
                contentPadding: EdgeInsets.zero,
                title: const Text('Category'),
                subtitle: Text(document.frontmatter.category),
              ),
              ListTile(
                contentPadding: EdgeInsets.zero,
                title: const Text('Date'),
                subtitle: Text(document.frontmatter.date),
              ),
              Text(
                'Content',
                style: Theme.of(context).textTheme.headline6,
              ),
              const SizedBox(
                height: 10,
              ),
              Expanded(
                child: MarkdownBody(
                  data: document.body,
                ),
              ),
            ],
          ),
        ),
      ),
    );
  }
}
```

`MarkdownBody` is a widget that displays markdown. It comes from the [flutter_markdown](https://pub.dev/packages/flutter_markdown) package. You can find more information about it [here](https://developer.school/tutorials/how-to-display-markdown-in-flutter).

CHANGELOG.md
## 1.0.3

Updated package versions.

## 1.0.2

* Added some minor documentation changes.

## 1.0.1

* Version updated in `README.md`
## 1.0.0

* Initial release


test/parser_test.dart
import 'package:cosmic_frontmatter/src/exceptions/frontmatter_parse_exception.dart';
import 'package:cosmic_frontmatter/src/functions/parse_frontmatter.dart';
import 'package:cosmic_frontmatter/src/model/mock_frontmatter.dart';
import 'package:flutter_test/flutter_test.dart';

import 'constants/test_constants.dart';

void main() {
  test('Parser should parse frontmatter', () {
    final mockFrontmatter = parseFrontmatter(
        content: mockDocument,
        frontmatterBuilder: (map) {
          // Do anything you want with the `map`:

          return MockFrontmatter.fromJson(map);
        });

    expect(mockFrontmatter.frontmatter.author, expectedMap['frontmatter']['author']);
  });
  test('Parser should not parse bad frontmatter', () {
    expect(
        () => parseFrontmatter(
            content: failureDocument,
            frontmatterBuilder: (map) {
              return MockFrontmatter.fromJson(map);
            }),
        throwsA(isA<FrontmatterParseException>()));
  });
}


test/document_test.dart
import 'package:cosmic_frontmatter/src/functions/parse_frontmatter.dart';
import 'package:cosmic_frontmatter/src/model/document.dart';
import 'package:cosmic_frontmatter/src/model/mock_frontmatter.dart';
import 'package:flutter_test/flutter_test.dart';

import 'constants/test_constants.dart';

void main() {
  test('Document can be parsed to json', () {
    final data = parseFrontmatter(
        content: mockDocument,
        frontmatterBuilder: (map) {
          return MockFrontmatter.fromJson(map);
        });

    expect(data.toJson((value) => value.toJson()), expectedMap);
  });
  test('Document can be parsed from json', () {
    final data =
        Document<MockFrontmatter>.fromJson(expectedMap, (d) => MockFrontmatter.fromJson((d as Map<String, dynamic>)));

    expect(data.frontmatter.toJson(), expectedMap['frontmatter']);
    expect(data.body, expectedMap['body']);
  });
}


test/constants/test_constants.dart
const Map<String, dynamic> expectedMap = {
  'frontmatter': {
    'title': 'This is a test document',
    'author': 'paul-halliday',
    'excerpt': 'Learn how to use Markdown with Flutter at developer.school',
    'category': 'Flutter',
    'date': '2021-10-25'
  },
  'body': 'Here is a test document with some data\n'
      '\n'
      'Other stuff'
};

const mockFrontmatter = """---
title: This is a test document
author: paul-halliday
excerpt: Learn how to use Markdown with Flutter at developer.school
category: Flutter
date: "2021-10-25"
---""";

const mockDocument = """
---
title: This is a test document
author: paul-halliday
excerpt: Learn how to use Markdown with Flutter at developer.school
category: Flutter
date: "2021-10-25"
---

Here is a test document with some data

Other stuff
""";

const failureDocument = """
title: This is a test document
author: paul-halliday
excerpt: Learn how to use Markdown with Flutter at developer.school
category: Flutter
date: "2021-10-25"

Here is a test document with some data

Other stuff
""";


lib/src/model/document.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'document.g.dart';

@JsonSerializable(
  genericArgumentFactories: true,
)

/// Representation of a Markdown document with frontmatter.
class Document<T> {
  final T frontmatter;
  final String body;

  const Document({required this.frontmatter, required this.body});

  factory Document.fromJson(
    Map<String, dynamic> json,
    T Function(Object? json) fromJsonT,
  ) =>
      _$DocumentFromJson(json, fromJsonT);

  Map<String, dynamic> toJson(Object Function(T value) toJsonT) => _$DocumentToJson(this, toJsonT);

  @override
  int get hashCode => Object.hash(
        frontmatter.hashCode,
        body,
      );

  @override
  bool operator ==(Object other) {
    return other is Document<T> && other.body == body && other.frontmatter.toString() == frontmatter.toString();
  }
}


lib/src/functions/parseFrontmatter.dart
import 'package:cosmic_frontmatter/src/exceptions/frontmatter_parse_exception.dart';
import 'package:cosmic_frontmatter/src/model/document.dart';
import 'package:yaml/yaml.dart';

/// Parses a YAML document into a [Document] object.
///
/// Throws a [FrontmatterParseException] if the document is not valid YAML.
Document<T> parseFrontmatter<T>({
  required String content,
  required T Function(Map<String, dynamic>) frontmatterBuilder,
  String delimiter = '---',
}) {
  final trimmedContent = content.trimLeft();

  if (!trimmedContent.startsWith(delimiter)) {
    throw const FrontmatterParseException.delimiter();
  }

  final delimiterEnd = trimmedContent.indexOf("\n$delimiter");
  final frontmatter = trimmedContent.substring(delimiter.length, delimiterEnd);
  final body = trimmedContent.substring(delimiterEnd + delimiter.length + 1).trim();

  final yamlAsMap = Map<String, dynamic>.from(loadYaml(frontmatter));
  final model = frontmatterBuilder(yamlAsMap);

  return Document(body: body, frontmatter: model);
}


lib/src/model/mock_frontmatter.freezed.dart
// coverage:ignore-file
// GENERATED CODE - DO NOT MODIFY BY HAND
// ignore_for_file: unused_element, deprecated_member_use, deprecated_member_use_from_same_package, use_function_type_syntax_for_parameters, unnecessary_const, avoid_init_to_null, invalid_override_different_default_values_named, prefer_expression_function_bodies, annotate_overrides, invalid_annotation_target

part of 'mock_frontmatter.dart';

// **************************************************************************
// FreezedGenerator
// **************************************************************************

T _$identity<T>(T value) => value;

final _privateConstructorUsedError = UnsupportedError(
    'It seems like you constructed your class using `MyClass._()`. This constructor is only meant to be used by freezed and you are not supposed to need it nor use it.\nPlease check the documentation here for more informations: https://github.com/rrousselGit/freezed#custom-getters-and-methods');

MockFrontmatter _$MockFrontmatterFromJson(Map<String, dynamic> json) {
  return _MockFrontmatter.fromJson(json);
}

/// @nodoc
class _$MockFrontmatterTearOff {
  const _$MockFrontmatterTearOff();

  _MockFrontmatter call(
      {required String title,
      required String author,
      required String excerpt,
      required String category,
      required String date}) {
    return _MockFrontmatter(
      title: title,
      author: author,
      excerpt: excerpt,
      category: category,
      date: date,
    );
  }

  MockFrontmatter fromJson(Map<String, Object?> json) {
    return MockFrontmatter.fromJson(json);
  }
}

/// @nodoc
const $MockFrontmatter = _$MockFrontmatterTearOff();

/// @nodoc
mixin _$MockFrontmatter {
  String get title => throw _privateConstructorUsedError;
  String get author => throw _privateConstructorUsedError;
  String get excerpt => throw _privateConstructorUsedError;
  String get category => throw _privateConstructorUsedError;
  String get date => throw _privateConstructorUsedError;

  Map<String, dynamic> toJson() => throw _privateConstructorUsedError;
  @JsonKey(ignore: true)
  $MockFrontmatterCopyWith<MockFrontmatter> get copyWith =>
      throw _privateConstructorUsedError;
}

/// @nodoc
abstract class $MockFrontmatterCopyWith<$Res> {
  factory $MockFrontmatterCopyWith(
          MockFrontmatter value, $Res Function(MockFrontmatter) then) =
      _$MockFrontmatterCopyWithImpl<$Res>;
  $Res call(
      {String title,
      String author,
      String excerpt,
      String category,
      String date});
}

/// @nodoc
class _$MockFrontmatterCopyWithImpl<$Res>
    implements $MockFrontmatterCopyWith<$Res> {
  _$MockFrontmatterCopyWithImpl(this._value, this._then);

  final MockFrontmatter _value;
  // ignore: unused_field
  final $Res Function(MockFrontmatter) _then;

  @override
  $Res call({
    Object? title = freezed,
    Object? author = freezed,
    Object? excerpt = freezed,
    Object? category = freezed,
    Object? date = freezed,
  }) {
    return _then(_value.copyWith(
      title: title == freezed
          ? _value.title
          : title // ignore: cast_nullable_to_non_nullable
              as String,
      author: author == freezed
          ? _value.author
          : author // ignore: cast_nullable_to_non_nullable
              as String,
      excerpt: excerpt == freezed
          ? _value.excerpt
          : excerpt // ignore: cast_nullable_to_non_nullable
              as String,
      category: category == freezed
          ? _value.category
          : category // ignore: cast_nullable_to_non_nullable
              as String,
      date: date == freezed
          ? _value.date
          : date // ignore: cast_nullable_to_non_nullable
              as String,
    ));
  }
}

/// @nodoc
abstract class _$MockFrontmatterCopyWith<$Res>
    implements $MockFrontmatterCopyWith<$Res> {
  factory _$MockFrontmatterCopyWith(
          _MockFrontmatter value, $Res Function(_MockFrontmatter) then) =
      __$MockFrontmatterCopyWithImpl<$Res>;
  @override
  $Res call(
      {String title,
      String author,
      String excerpt,
      String category,
      String date});
}

/// @nodoc
class __$MockFrontmatterCopyWithImpl<$Res>
    extends _$MockFrontmatterCopyWithImpl<$Res>
    implements _$MockFrontmatterCopyWith<$Res> {
  __$MockFrontmatterCopyWithImpl(
      _MockFrontmatter _value, $Res Function(_MockFrontmatter) _then)
      : super(_value, (v) => _then(v as _MockFrontmatter));

  @override
  _MockFrontmatter get _value => super._value as _MockFrontmatter;

  @override
  $Res call({
    Object? title = freezed,
    Object? author = freezed,
    Object? excerpt = freezed,
    Object? category = freezed,
    Object? date = freezed,
  }) {
    return _then(_MockFrontmatter(
      title: title == freezed
          ? _value.title
          : title // ignore: cast_nullable_to_non_nullable
              as String,
      author: author == freezed
          ? _value.author
          : author // ignore: cast_nullable_to_non_nullable
              as String,
      excerpt: excerpt == freezed
          ? _value.excerpt
          : excerpt // ignore: cast_nullable_to_non_nullable
              as String,
      category: category == freezed
          ? _value.category
          : category // ignore: cast_nullable_to_non_nullable
              as String,
      date: date == freezed
          ? _value.date
          : date // ignore: cast_nullable_to_non_nullable
              as String,
    ));
  }
}

/// @nodoc
@JsonSerializable()
class _$_MockFrontmatter extends _MockFrontmatter {
  const _$_MockFrontmatter(
      {required this.title,
      required this.author,
      required this.excerpt,
      required this.category,
      required this.date})
      : super._();

  factory _$_MockFrontmatter.fromJson(Map<String, dynamic> json) =>
      _$$_MockFrontmatterFromJson(json);

  @override
  final String title;
  @override
  final String author;
  @override
  final String excerpt;
  @override
  final String category;
  @override
  final String date;

  @override
  String toString() {
    return 'MockFrontmatter(title: $title, author: $author, excerpt: $excerpt, category: $category, date: $date)';
  }

  @override
  bool operator ==(dynamic other) {
    return identical(this, other) ||
        (other.runtimeType == runtimeType &&
            other is _MockFrontmatter &&
            (identical(other.title, title) || other.title == title) &&
            (identical(other.author, author) || other.author == author) &&
            (identical(other.excerpt, excerpt) || other.excerpt == excerpt) &&
            (identical(other.category, category) ||
                other.category == category) &&
            (identical(other.date, date) || other.date == date));
  }

  @override
  int get hashCode =>
      Object.hash(runtimeType, title, author, excerpt, category, date);

  @JsonKey(ignore: true)
  @override
  _$MockFrontmatterCopyWith<_MockFrontmatter> get copyWith =>
      __$MockFrontmatterCopyWithImpl<_MockFrontmatter>(this, _$identity);

  @override
  Map<String, dynamic> toJson() {
    return _$$_MockFrontmatterToJson(this);
  }
}

abstract class _MockFrontmatter extends MockFrontmatter {
  const factory _MockFrontmatter(
      {required String title,
      required String author,
      required String excerpt,
      required String category,
      required String date}) = _$_MockFrontmatter;
  const _MockFrontmatter._() : super._();

  factory _MockFrontmatter.fromJson(Map<String, dynamic> json) =
      _$_MockFrontmatter.fromJson;

  @override
  String get title;
  @override
  String get author;
  @override
  String get excerpt;
  @override
  String get category;
  @override
  String get date;
  @override
  @JsonKey(ignore: true)
  _$MockFrontmatterCopyWith<_MockFrontmatter> get copyWith =>
      throw _privateConstructorUsedError;
}


lib/src/exceptions/frontmatter_parse_exception.freezed.dart
// coverage:ignore-file
// GENERATED CODE - DO NOT MODIFY BY HAND
// ignore_for_file: unused_element, deprecated_member_use, deprecated_member_use_from_same_package, use_function_type_syntax_for_parameters, unnecessary_const, avoid_init_to_null, invalid_override_different_default_values_named, prefer_expression_function_bodies, annotate_overrides, invalid_annotation_target

part of 'frontmatter_parse_exception.dart';

// **************************************************************************
// FreezedGenerator
// **************************************************************************

T _$identity<T>(T value) => value;

final _privateConstructorUsedError = UnsupportedError(
    'It seems like you constructed your class using `MyClass._()`. This constructor is only meant to be used by freezed and you are not supposed to need it nor use it.\nPlease check the documentation here for more informations: https://github.com/rrousselGit/freezed#custom-getters-and-methods');

/// @nodoc
class _$FrontmatterParseExceptionTearOff {
  const _$FrontmatterParseExceptionTearOff();

  _FrontmatterParseException call(String message) {
    return _FrontmatterParseException(
      message,
    );
  }

  FrontmatterParseExceptionDelimiter delimiter(
      [String message =
          'Frontmatter document must start with your delimiter.']) {
    return FrontmatterParseExceptionDelimiter(
      message,
    );
  }
}

/// @nodoc
const $FrontmatterParseException = _$FrontmatterParseExceptionTearOff();

/// @nodoc
mixin _$FrontmatterParseException {
  String get message => throw _privateConstructorUsedError;

  @optionalTypeArgs
  TResult when<TResult extends Object?>(
    TResult Function(String message) $default, {
    required TResult Function(String message) delimiter,
  }) =>
      throw _privateConstructorUsedError;
  @optionalTypeArgs
  TResult? whenOrNull<TResult extends Object?>(
    TResult Function(String message)? $default, {
    TResult Function(String message)? delimiter,
  }) =>
      throw _privateConstructorUsedError;
  @optionalTypeArgs
  TResult maybeWhen<TResult extends Object?>(
    TResult Function(String message)? $default, {
    TResult Function(String message)? delimiter,
    required TResult orElse(),
  }) =>
      throw _privateConstructorUsedError;
  @optionalTypeArgs
  TResult map<TResult extends Object?>(
    TResult Function(_FrontmatterParseException value) $default, {
    required TResult Function(FrontmatterParseExceptionDelimiter value)
        delimiter,
  }) =>
      throw _privateConstructorUsedError;
  @optionalTypeArgs
  TResult? mapOrNull<TResult extends Object?>(
    TResult Function(_FrontmatterParseException value)? $default, {
    TResult Function(FrontmatterParseExceptionDelimiter value)? delimiter,
  }) =>
      throw _privateConstructorUsedError;
  @optionalTypeArgs
  TResult maybeMap<TResult extends Object?>(
    TResult Function(_FrontmatterParseException value)? $default, {
    TResult Function(FrontmatterParseExceptionDelimiter value)? delimiter,
    required TResult orElse(),
  }) =>
      throw _privateConstructorUsedError;

  @JsonKey(ignore: true)
  $FrontmatterParseExceptionCopyWith<FrontmatterParseException> get copyWith =>
      throw _privateConstructorUsedError;
}

/// @nodoc
abstract class $FrontmatterParseExceptionCopyWith<$Res> {
  factory $FrontmatterParseExceptionCopyWith(FrontmatterParseException value,
          $Res Function(FrontmatterParseException) then) =
      _$FrontmatterParseExceptionCopyWithImpl<$Res>;
  $Res call({String message});
}

/// @nodoc
class _$FrontmatterParseExceptionCopyWithImpl<$Res>
    implements $FrontmatterParseExceptionCopyWith<$Res> {
  _$FrontmatterParseExceptionCopyWithImpl(this._value, this._then);

  final FrontmatterParseException _value;
  // ignore: unused_field
  final $Res Function(FrontmatterParseException) _then;

  @override
  $Res call({
    Object? message = freezed,
  }) {
    return _then(_value.copyWith(
      message: message == freezed
          ? _value.message
          : message // ignore: cast_nullable_to_non_nullable
              as String,
    ));
  }
}

/// @nodoc
abstract class _$FrontmatterParseExceptionCopyWith<$Res>
    implements $FrontmatterParseExceptionCopyWith<$Res> {
  factory _$FrontmatterParseExceptionCopyWith(_FrontmatterParseException value,
          $Res Function(_FrontmatterParseException) then) =
      __$FrontmatterParseExceptionCopyWithImpl<$Res>;
  @override
  $Res call({String message});
}

/// @nodoc
class __$FrontmatterParseExceptionCopyWithImpl<$Res>
    extends _$FrontmatterParseExceptionCopyWithImpl<$Res>
    implements _$FrontmatterParseExceptionCopyWith<$Res> {
  __$FrontmatterParseExceptionCopyWithImpl(_FrontmatterParseException _value,
      $Res Function(_FrontmatterParseException) _then)
      : super(_value, (v) => _then(v as _FrontmatterParseException));

  @override
  _FrontmatterParseException get _value =>
      super._value as _FrontmatterParseException;

  @override
  $Res call({
    Object? message = freezed,
  }) {
    return _then(_FrontmatterParseException(
      message == freezed
          ? _value.message
          : message // ignore: cast_nullable_to_non_nullable
              as String,
    ));
  }
}

/// @nodoc

class _$_FrontmatterParseException extends _FrontmatterParseException {
  const _$_FrontmatterParseException(this.message) : super._();

  @override
  final String message;

  @override
  String toString() {
    return 'FrontmatterParseException(message: $message)';
  }

  @override
  bool operator ==(dynamic other) {
    return identical(this, other) ||
        (other.runtimeType == runtimeType &&
            other is _FrontmatterParseException &&
            (identical(other.message, message) || other.message == message));
  }

  @override
  int get hashCode => Object.hash(runtimeType, message);

  @JsonKey(ignore: true)
  @override
  _$FrontmatterParseExceptionCopyWith<_FrontmatterParseException>
      get copyWith =>
          __$FrontmatterParseExceptionCopyWithImpl<_FrontmatterParseException>(
              this, _$identity);

  @override
  @optionalTypeArgs
  TResult when<TResult extends Object?>(
    TResult Function(String message) $default, {
    required TResult Function(String message) delimiter,
  }) {
    return $default(message);
  }

  @override
  @optionalTypeArgs
  TResult? whenOrNull<TResult extends Object?>(
    TResult Function(String message)? $default, {
    TResult Function(String message)? delimiter,
  }) {
    return $default?.call(message);
  }

  @override
  @optionalTypeArgs
  TResult maybeWhen<TResult extends Object?>(
    TResult Function(String message)? $default, {
    TResult Function(String message)? delimiter,
    required TResult orElse(),
  }) {
    if ($default != null) {
      return $default(message);
    }
    return orElse();
  }

  @override
  @optionalTypeArgs
  TResult map<TResult extends Object?>(
    TResult Function(_FrontmatterParseException value) $default, {
    required TResult Function(FrontmatterParseExceptionDelimiter value)
        delimiter,
  }) {
    return $default(this);
  }

  @override
  @optionalTypeArgs
  TResult? mapOrNull<TResult extends Object?>(
    TResult Function(_FrontmatterParseException value)? $default, {
    TResult Function(FrontmatterParseExceptionDelimiter value)? delimiter,
  }) {
    return $default?.call(this);
  }

  @override
  @optionalTypeArgs
  TResult maybeMap<TResult extends Object?>(
    TResult Function(_FrontmatterParseException value)? $default, {
    TResult Function(FrontmatterParseExceptionDelimiter value)? delimiter,
    required TResult orElse(),
  }) {
    if ($default != null) {
      return $default(this);
    }
    return orElse();
  }
}

abstract class _FrontmatterParseException extends FrontmatterParseException {
  const factory _FrontmatterParseException(String message) =
      _$_FrontmatterParseException;
  const _FrontmatterParseException._() : super._();

  @override
  String get message;
  @override
  @JsonKey(ignore: true)
  _$FrontmatterParseExceptionCopyWith<_FrontmatterParseException>
      get copyWith => throw _privateConstructorUsedError;
}

/// @nodoc
abstract class $FrontmatterParseExceptionDelimiterCopyWith<$Res>
    implements $FrontmatterParseExceptionCopyWith<$Res> {
  factory $FrontmatterParseExceptionDelimiterCopyWith(
          FrontmatterParseExceptionDelimiter value,
          $Res Function(FrontmatterParseExceptionDelimiter) then) =
      _$FrontmatterParseExceptionDelimiterCopyWithImpl<$Res>;
  @override
  $Res call({String message});
}

/// @nodoc
class _$FrontmatterParseExceptionDelimiterCopyWithImpl<$Res>
    extends _$FrontmatterParseExceptionCopyWithImpl<$Res>
    implements $FrontmatterParseExceptionDelimiterCopyWith<$Res> {
  _$FrontmatterParseExceptionDelimiterCopyWithImpl(
      FrontmatterParseExceptionDelimiter _value,
      $Res Function(FrontmatterParseExceptionDelimiter) _then)
      : super(_value, (v) => _then(v as FrontmatterParseExceptionDelimiter));

  @override
  FrontmatterParseExceptionDelimiter get _value =>
      super._value as FrontmatterParseExceptionDelimiter;

  @override
  $Res call({
    Object? message = freezed,
  }) {
    return _then(FrontmatterParseExceptionDelimiter(
      message == freezed
          ? _value.message
          : message // ignore: cast_nullable_to_non_nullable
              as String,
    ));
  }
}

/// @nodoc

class _$FrontmatterParseExceptionDelimiter
    extends FrontmatterParseExceptionDelimiter {
  const _$FrontmatterParseExceptionDelimiter(
      [this.message = 'Frontmatter document must start with your delimiter.'])
      : super._();

  @JsonKey(defaultValue: 'Frontmatter document must start with your delimiter.')
  @override
  final String message;

  @override
  String toString() {
    return 'FrontmatterParseException.delimiter(message: $message)';
  }

  @override
  bool operator ==(dynamic other) {
    return identical(this, other) ||
        (other.runtimeType == runtimeType &&
            other is FrontmatterParseExceptionDelimiter &&
            (identical(other.message, message) || other.message == message));
  }

  @override
  int get hashCode => Object.hash(runtimeType, message);

  @JsonKey(ignore: true)
  @override
  $FrontmatterParseExceptionDelimiterCopyWith<
          FrontmatterParseExceptionDelimiter>
      get copyWith => _$FrontmatterParseExceptionDelimiterCopyWithImpl<
          FrontmatterParseExceptionDelimiter>(this, _$identity);

  @override
  @optionalTypeArgs
  TResult when<TResult extends Object?>(
    TResult Function(String message) $default, {
    required TResult Function(String message) delimiter,
  }) {
    return delimiter(message);
  }

  @override
  @optionalTypeArgs
  TResult? whenOrNull<TResult extends Object?>(
    TResult Function(String message)? $default, {
    TResult Function(String message)? delimiter,
  }) {
    return delimiter?.call(message);
  }

  @override
  @optionalTypeArgs
  TResult maybeWhen<TResult extends Object?>(
    TResult Function(String message)? $default, {
    TResult Function(String message)? delimiter,
    required TResult orElse(),
  }) {
    if (delimiter != null) {
      return delimiter(message);
    }
    return orElse();
  }

  @override
  @optionalTypeArgs
  TResult map<TResult extends Object?>(
    TResult Function(_FrontmatterParseException value) $default, {
    required TResult Function(FrontmatterParseExceptionDelimiter value)
        delimiter,
  }) {
    return delimiter(this);
  }

  @override
  @optionalTypeArgs
  TResult? mapOrNull<TResult extends Object?>(
    TResult Function(_FrontmatterParseException value)? $default, {
    TResult Function(FrontmatterParseExceptionDelimiter value)? delimiter,
  }) {
    return delimiter?.call(this);
  }

  @override
  @optionalTypeArgs
  TResult maybeMap<TResult extends Object?>(
    TResult Function(_FrontmatterParseException value)? $default, {
    TResult Function(FrontmatterParseExceptionDelimiter value)? delimiter,
    required TResult orElse(),
  }) {
    if (delimiter != null) {
      return delimiter(this);
    }
    return orElse();
  }
}

abstract class FrontmatterParseExceptionDelimiter
    extends FrontmatterParseException {
  const factory FrontmatterParseExceptionDelimiter([String message]) =
      _$FrontmatterParseExceptionDelimiter;
  const FrontmatterParseExceptionDelimiter._() : super._();

  @override
  String get message;
  @override
  @JsonKey(ignore: true)
  $FrontmatterParseExceptionDelimiterCopyWith<
          FrontmatterParseExceptionDelimiter>
      get copyWith => throw _privateConstructorUsedError;
}


example/lib/main.dart
import 'package:cosmic_frontmatter/cosmic_frontmatter.dart';
import 'package:example/models/mock_frontmatter.dart';
import 'package:flutter/material.dart';
import 'package:flutter_markdown/flutter_markdown.dart';

const mockDocument = """
---
title: This is a test document
author: paul-halliday
excerpt: Learn how to use Markdown with Flutter at developer.school
category: Flutter
date: "2021-10-25"
---

You can learn how to use Markdown with Flutter at [developer.school](https://developer.school/tutorials/how-to-display-markdown-in-flutter).
""";

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'cosmic_frontmatter',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key}) : super(key: key);
  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  late Document<MockFrontmatter> document;

  @override
  void initState() {
    super.initState();

    document = parseFrontmatter(
      content: mockDocument,
      frontmatterBuilder: (map) => MockFrontmatter.fromJson(map),
    );
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(document.frontmatter.title),
      ),
      body: Padding(
        padding: const EdgeInsets.all(16.0),
        child: Center(
          child: Column(
            crossAxisAlignment: CrossAxisAlignment.stretch,
            children: [
              ListTile(
                contentPadding: EdgeInsets.zero,
                title: const Text('Title'),
                subtitle: Text(document.frontmatter.title),
              ),
              ListTile(
                contentPadding: EdgeInsets.zero,
                title: const Text('Author'),
                subtitle: Text(document.frontmatter.author),
              ),
              ListTile(
                contentPadding: EdgeInsets.zero,
                title: const Text('Excerpt'),
                subtitle: Text(document.frontmatter.excerpt),
              ),
              ListTile(
                contentPadding: EdgeInsets.zero,
                title: const Text('Category'),
                subtitle: Text(document.frontmatter.category),
              ),
              ListTile(
                contentPadding: EdgeInsets.zero,
                title: const Text('Date'),
                subtitle: Text(document.frontmatter.date),
              ),
              Text(
                'Content',
                style: Theme.of(context).textTheme.headline6,
              ),
              const SizedBox(
                height: 10,
              ),
              Expanded(
                child: MarkdownBody(
                  data: document.body,
                ),
              ),
            ],
          ),
        ),
      ),
    );
  }
}
